pkg/check/node_perf.go

```diff
@@ -0,0 +1,197 @@
+		},
+	}
+)
+
+func (c *CheckNodeDiskPerf) BuildCounterIdMap(ctx *CheckContext, vm *mo.VirtualMachine) error {
+	buildCounterIDMapLock.Lock()
+	defer buildCounterIDMapLock.Unlock()
+
+	if len(perfMetricIDMap) > 0 {
+		return nil
+	}
+	var counterIds []int32
+
+	perf := types.QueryAvailablePerfMetric{
+		This:   *ctx.VMClient.ServiceContent.PerfManager,
+		Entity: vm.ManagedEntity.Reference(),
+	}
+
+	perfList, err := methods.QueryAvailablePerfMetric(ctx.Context, ctx.VMClient.RoundTripper, &perf)
+	if err != nil {
+		klog.V(2).Infof("error getting available performance metrics: %s", err.Error())
+		return err
+	}
+
+	var counterIDPerfMetricIDMap = map[int32]*types.PerfMetricId{}
+	klog.V(4).Infof("found [%d] performance metrics available", len(perfList.Returnval))
+	for _, perf := range perfList.Returnval {
+		counterIds = append(counterIds, perf.CounterId)
+		copyPerf := perf
+		counterIDPerfMetricIDMap[perf.CounterId] = &copyPerf
+	}
+
+	perfCounter := types.QueryPerfCounter{
+		This:      *ctx.VMClient.ServiceContent.PerfManager,
+		CounterId: counterIds,
+	}
+	perfCounters, err := methods.QueryPerfCounter(ctx.Context, ctx.VMClient.RoundTripper, &perfCounter)
+	if err != nil {
+		return err
+	}
+
+	for _, perfCounter := range perfCounters.Returnval {
+		if _, ok := counterIDPerfMetricIDMap[perfCounter.Key]; !ok {
+			continue
+		}
+		perfMetricIDMap[perfCounter.Name()] = counterIDPerfMetricIDMap[perfCounter.Key]
+	}
+	return nil
+}
+
+func (c *CheckNodeDiskPerf) GetPerfMetric(ctx *CheckContext, vm *mo.VirtualMachine, metricName string) (*types.PerfMetricId, error) {
+	err := c.BuildCounterIdMap(ctx, vm)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := perfMetricIDMap[metricName]; !ok {
+		return nil, fmt.Errorf("unable to find metric with name[%s]", metricName)
+	}
+	return perfMetricIDMap[metricName], nil
+}
+
+func (c *CheckNodeDiskPerf) StartCheck() error {
+	return nil
+}
+
+func (c *CheckNodeDiskPerf) PerformMetricCheck(ctx *CheckContext, node *v1.Node, vm *mo.VirtualMachine, mcd MetricCheckDef) error {
+	var (
+		hasWarning = false
+		hasError   = false
+	)
+
+	_, isMaster := node.Labels["node-role.kubernetes.io/master"]
+	_, isWorker := node.Labels["node-role.kubernetes.io/worker"]
+
+	if isMaster && mcd.checkMasters == false {
+		klog.V(2).Infof("skipping metric check of[%s] as it is not intended for master nodes", mcd.metricName)
+		return nil
+	} else if isWorker && mcd.checkWorkers == false {
+		if isMaster && mcd.checkMasters {
+			klog.V(2).Infof("master nodes are schedulable, allowing check of [%s] even though check is disabled for workers", mcd.metricName)
+		} else {
+			klog.V(2).Infof("skipping metric check of[%s] as it is not intended for worker nodes", mcd.metricName)
+			return nil
+		}
+
+	}
+
+	klog.V(2).Infof("performing metric check of: %s", mcd.metricName)
+
+	perfMetricID, err := c.GetPerfMetric(ctx, vm, mcd.metricName)
+	if err != nil {
+		klog.V(2).Infof("error getting perf metric id: %s", err.Error())
```

Comment:
Why is the error thrown away? A comment would be helpful.

Reply:
I was concerned about polluting the check results with that error.  But I guess it would make sense to propagate that back up to the caller as a failure there would be something worth noting in the results.